src/datascience-ui/data-explorer/data-wrangler/mainPanel.tsx

```diff
@@ -0,0 +1,632 @@
+            this.sentDone = true;
+            this.sendMessage(DataWranglerMessages.CompletedData);
+        }
+
+        const progressBar = this.state.totalRowCount > this.state.fetchedRowCount ? <Progress /> : undefined;
+
+        return (
+            <div className="main-panel" ref={this.container}>
+                <StyleInjector
+                    onReady={this.saveReadyState}
+                    settings={this.state.settings}
+                    expectingDark={this.props.baseTheme !== 'vscode-light'}
+                    postOffice={this.postOffice}
+                />
+                {progressBar}
+                {/* {this.renderBreadcrumb()} */}
+                <Toolbar
+                    handleRefreshRequest={this.handleRefreshRequest}
+                    submitCommand={this.submitCommand}
+                    onToggleFilter={() => this.toggleFilterEvent.notify()}
+                />
+                {this.renderSliceControls()}
+                {this.state.styleReady && this.renderGrid()}
+            </div>
+        );
+    };
+
+    public renderSliceControls = () => {
+        if (
+            this.state.isSliceDataEnabled &&
+            this.state.originalVariableShape &&
+            this.state.originalVariableShape.filter((v) => !!v).length > 1
+        ) {
+            return (
+                <SliceControl
+                    sliceExpression={this.state.sliceExpression}
+                    loadingData={this.state.totalRowCount > this.state.fetchedRowCount}
+                    originalVariableShape={this.state.originalVariableShape}
+                    handleSliceRequest={this.handleSliceRequest}
+                    onCheckboxToggled={this.handleCheckboxToggle}
+                    onPanelToggled={() => this.resizeGridEvent.notify()}
+                />
+            );
+        }
+    };
+
+    // private renderBreadcrumb() {
+    //     let breadcrumbText = this.state.variableName;
+    //     if (this.state.originalVariableShape) {
+    //         breadcrumbText += ' (' + this.state.originalVariableShape?.join(', ') + ')';
+    //     }
+    //     if (breadcrumbText) {
+    //         return (
+    //             <div className="breadcrumb-container control-container">
+    //                 <div className="breadcrumb">
+    //                     <div className="icon-python breadcrumb-file-icon" />
+    //                     <span>{this.state.fileName}</span>
+    //                     {this.state.fileName ? (
+    //                         <div className="codicon codicon-chevron-right breadcrumb-codicon" />
+    //                     ) : undefined}
+    //                     <span>{breadcrumbText}</span>
+    //                 </div>
+    //             </div>
+    //         );
+    //     }
+    // }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    public handleMessage = (msg: string, payload?: any) => {
```

Comment:
I wonder if we can abstract out part of this. A lot of the webviews have to handle the onigasm stuff. Seems like we don't need to repeat it in each.

Reply:
The state transitions in the datawrangler may warrant using redux instead of handling everything in the react code, but I think we can leave that till later/next pr.